### DISCUSSION TITLE
如何使用其他项目的类静态成员

### DISCUSSION THREAD

**Discussion by tsctsc6** [BYJRK/DotNet-Discussions#315] [C# 技术交流]
比如我最近创建了avalnia.xplat项目，名称姑且叫AvalniaApp1吧。

它的项目结构是AvalniaApp1.Desktop，引用了AvalniaApp1。如果生成桌面端应用，那么程序的入口在AvalniaApp1.Desktop的Program.cs中。

现在我希望做一个全局异常捕抓并记录日志（日志库使用SeriLog，使用静态类Log.logger注册logger），官方文档的建议是，在AvalniaApp1.Desktop的Program.cs中的main函数进行try catch，这并没有问题。问题是我在AvalniaApp1项目中如何使用AvalniaApp1.Desktop的logger?

我尝试既在AvalniaApp1.Desktop中新建一个log实例，又在AvalniaApp1中新建一个log实例。但是这样会造成冲突，导致有多个日志文件产生，日志被分裂到不同的日志文件中。

**Comment by CodingOctocat** [comment]
你可能需要一个中间项目做桥梁，比如：AvalniaApp1.Logger，其他两个项目引用它。

**Reply by tsctsc6** [reply]
照你这么说，两个项目都引用了SeriLog这个库，就相当于中间项目咯

**Reply by arcanix601** [reply]
是这样的，你可直接用Serilog中的静态类Log，只要他在入口项目注册好了就行

**Comment by tsctsc6** [comment]
天啊，只要在AvalniaApp1.Desktop中注册了Log.logger这个静态成员，本来就可以跨项目访问了。😂此贴结束

**Reply by BYJRK** [reply]
确实如此，但是大一些的项目，这并不是一个好方法，因为你很可能还需要使用 IoC 容器。在那种情况下，推荐参考微软自己的 DI 的做法，即写对于 `ServiceCollection` 的扩展方法（通常名称为 `AddXXX`），从而实现多项目的 IoC 服务注册。

**Reply by CodingOctocat** [reply]
不是吧，你的 Desktop 是 WPF 项目，其他项目能正确引用？一般来说需要一个类库项目作为中间项目啊。

**Reply by tsctsc6** [reply]
？

AvalniaApp1.Desktop项目，引用了AvalniaApp1项目。然后这两个项目都引用了SeriLog库。SeriLog库中有静态成员Log.logger。

**Comment by tsctsc6** [comment]
我只是之前脑子抽了，以为AvalniaApp1.Desktop项目和AvalniaApp1项目中的Log.logger不是同一个东西。
